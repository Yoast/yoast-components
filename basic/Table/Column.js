import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";
import colors from "../../style-guide/colors.json";

/**
 * Separatifies an element.
 *
 * @returns {String} CSS
 */
export function separatify() {
	return `
		&::after {
			position:relative;
			display: inline-block;
			border-right: 2px solid ${ colors.$color_grey };
			padding-right: 40px;
			height: 60px;
			content: "";
		}

		@media screen and ( max-width: 1355px ) {
			&::after {
				padding-right: 24px;
			}
		}

		@media screen and ( max-width: 800px ) {
			&::after {
				height: 48px;
				padding-right: 16px;
			}
		}
	`;
}

/**
 * Make an element text nowrap and add ellipsis.
 *
 * @returns {String} CSS
 */
export function ellipsify() {
	return `
		overflow: hidden;
		text-overflow: ellipsis;
		white-space: nowrap;
	`;
}

/*
 * The "Columns" are flex-items, their "flex" property is the initial `0 1 auto`
 * that translates to: cannot grow, can shrink, initial width based on the content.
 * A column can be changed on a case by case basis wrapping the Columns in a
 * styled component.
 */
const ColumnBase = styled.span`
	font-size: 14px;
	padding-left: 40px;

	&:first-child {
		padding-left: 0;
	}

	&::before {
 		position: absolute;
 		left: -9999em;
 		top: -30px;
 		font-size: 1.286em;
 		line-height: 0;
 		${ props => props.headerLabel ? `content: "${props.headerLabel}";` : "content: none;" }
 	}

 	${ props => props.separator ? separatify() : "" }
	${ props => props.ellipsis ? ellipsify() : "" }

	@media screen and ( max-width: 1355px ) {
		padding-left: 20px;
		${ props => props.hideOnTablet ? "display: none;" : "" }
	}

	@media screen and ( max-width: 800px ) {
		${ props => props.hideOnMobile ? "display: none;" : "" }
	}
`;

ColumnBase.propTypes = {
	children: PropTypes.any,
	hideOnMobile: PropTypes.bool,
	hideOnTablet: PropTypes.bool,
	separator: PropTypes.bool,
	headerLabel: PropTypes.string,
	ellipsis: PropTypes.bool,
};

ColumnBase.defaultProps = {
	hideOnMobile: false,
	hideOnTable: false,
	separator: false,
	ellipsis: false,
};

/*
 * Primary column, the largest one in a row: can grow, cannot shrink, and the
 * initial width is 200 pixels. In the responsive view, can shrink.
 */
export const ColumnPrimary = styled( ColumnBase )`
	flex: 1 0 200px;

	@media screen and ( max-width: 800px ) {
		flex-shrink: 1;
	}
`;

/*
 * Column with fixed width: cannot grow, cannot shrink, and the width based on
 * its content.
 */
export const ColumnFixedWidth = styled( ColumnBase )`
	flex: 0 0 auto;
`;

/*
 * Column with a minimum width: can grow, cannot shrink, and the initial width
 * is 100 pixels.
 */
export const ColumnMinWidth = styled( ColumnBase )`
	flex: 1 0 100px;
`;

/*
 * Column with icon: cannot grow, cannot shrink, and the width is fixed based on
 * its content. The height is smaller in the responsive view because icons are
 * smaller.
 */
export const ColumnIcon = styled( ColumnFixedWidth )`
	height: 60px;

	@media screen and ( max-width: 800px ) {
		height: 48px;
	}
`;

/**
 * Makes columns use headers as bold labels with a colon in the mobile responsive view.
 *
 * @param {ReactElement} column The original column.
 *
 * @returns {ReactElement} The column with transformed headers.
 */
export function responsiveHeaders( column ) {
	return styled( column )`
		@media screen and ( max-width: 800px ) {
			&::before {
				float: left;
				min-width: 60px;
				line-height: inherit;
				font-weight: 700;
				${ props => props.headerLabel ? `content: "${props.headerLabel}:";` : "content: none;" }
			}

			> span {
				line-height: inherit;
			}
		}
	`;
}
