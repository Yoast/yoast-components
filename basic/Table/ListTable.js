import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";
import colors from "../../style-guide/colors.json";

const List = styled.ul`
	margin: 0;
 	padding: 0;
 	list-style: none;
 	position: relative;
 	width: 100%;

 	li:first-child {
		& > span::before {
			left: auto;
		}
	}
`;

List.propTypes = {
	children: PropTypes.any,
};

/**
 * Makes an element full-width in the mobile responsive view.
 *
 * @param {ReactElement} component The original element.
 * @returns {ReactElement} The element with full width responsive style.
 */
export function makeFullWidth( component ) {
	return styled( component )`
		@media screen and ( max-width: 800px ) {
			min-width: 100%;
			margin-top: 1em;
			padding-right: 0;
			padding-left: 0;
		}
	`;
}

class ListTable extends React.Component {
	constructor( props ) {
		super( props );

		this.children = this.getChildren();
	}

	getChildren() {
		if ( this.props.children === 1 ) {
			return [ this.props.children ];
		}

		return this.props.children;
	}

	render() {
		return ( <List role="list" children={ this.children } /> );
	}
}

class ZebrafiedTable extends ListTable {
	constructor( props ) {
		super( props );

		this.zebraProps = Object.assign( {}, props );
		this.zebrafyChildren();
	}

	zebrafyChildren() {
		this.zebraProps.children = this.props.children.map( ( child, index ) => {
			return React.cloneElement( child, {
				background: ( index % 2 === 1 ) ? colors.$color_white : colors.$color_background_light,
			} );
		} );
	}

	render() {
		return ( <List role="list" { ...this.zebraProps } /> );
	}
}

ListTable.propTypes = {
	children: PropTypes.oneOfType( [
		PropTypes.arrayOf( PropTypes.node ),
		PropTypes.node,
	] )
};

export { ListTable, ZebrafiedTable };
