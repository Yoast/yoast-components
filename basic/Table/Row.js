import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";
import colors from "../../style-guide/colors.json";

// The Rows are flex containers.
export const Row = styled.li`
	background: ${ props => props.background };
	display: flex;
	align-items: center;
	justify-content: space-between;
`;

Row.propTypes = {
	background: PropTypes.string,
	hasHeaderLabels: PropTypes.bool,
};

Row.defaultProps = {
	background: colors.$color_white,
	hasHeaderLabels: true,
};

/*
 * A responsive row allows children to wrap in new lines in the responsive view.
 * If the columns have headers, they're displayed as labels inside the column
 * content.
 */
export const RowMobileCollapse = styled( Row )`
	@media screen and ( max-width: 800px ) {
		flex-wrap: wrap;
		align-items: flex-start;

		&:first-child {
			margin-top: ${ props => props.hasHeaderLabels ? "24px" : "0" };
		}

		// Use the column headers (if any) as labels.
		& > span::before {
			position: static;
			display: inline-block;
			padding-right: 0.5em;
			font-size: inherit;
		}
	}
`;
