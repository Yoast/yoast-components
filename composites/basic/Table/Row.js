import PropTypes from "prop-types";
import styled from "styled-components";
import colors from "../../../style-guide/colors.json";

// The Rows are flex containers.
export const Row = styled.li`
	background: ${ props => props.background };
	display: flex;
	min-height: 100px;
	align-items: center;
	justify-content: space-between;
`;

Row.propTypes = {
	background: PropTypes.string,
	hasHeaderLabels: PropTypes.bool,
};

Row.defaultProps = {
	background: colors.$color_white,
	hasHeaderLabels: true,
};

/*
 * A row with responsive wrap allows children to wrap in new lines in the responsive view.
 * If the columns have headers, they're displayed as labels inside the column
 * content.
 */
export const RowResponsiveWrap = styled( Row )`
	@media screen and ( max-width: 800px ) {
		flex-wrap: wrap;
		align-items: flex-start;

		&:first-child {
			margin-top: ${ props => props.hasHeaderLabels ? "24px" : "0" };
		}

		// Use the column headers (if any) as labels.
		& > span::before {
			position: static;
			display: inline-block;
			padding-right: 0.5em;
			font-size: inherit;
		}
		& > span {
			padding-left: 0;
		}
	}
`;
