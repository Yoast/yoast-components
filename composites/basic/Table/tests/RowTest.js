import React from "react";
import renderer from "react-test-renderer";

import { Row, RowResponsiveWrap } from "../../Table/Row";

test( "the Row component matches the snapshot", () => {
	const component = renderer.create(
		<Row />
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "the RowResponsiveWrap component matches the snapshot", () => {
	const component = renderer.create(
		<RowResponsiveWrap />
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

