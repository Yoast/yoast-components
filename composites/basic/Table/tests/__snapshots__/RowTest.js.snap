// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Row component matches the snapshot 1`] = `
<li
  className="sc-bdVaJa gRzYbk"
/>
`;

exports[`the RowResponsiveWrap component matches the snapshot 1`] = `
<li
  className="sc-bwzfXH hEqAjG sc-bdVaJa gRzYbk"
/>
`;
